reference-implementation/lib/helpers/webidl.js

```diff
@@ -0,0 +1,203 @@
+'use strict';
+const utils = require('../../generated/utils.js');
+const { rethrowAssertionErrorRejection } = require('./miscellaneous.js');
+
+// https://heycam.github.io/webidl/#new
+// https://github.com/jsdom/webidl2js/issues/193
+exports.webidlNew = (globalObject, typeName, implModule) => {
+  if (globalObject[utils.ctorRegistrySymbol] === undefined) {
+    throw new Error('Internal error: invalid global object');
+  }
+
+  const ctor = globalObject[utils.ctorRegistrySymbol][typeName];
+  if (ctor === undefined) {
+    throw new Error(`Internal error: constructor ${typeName} is not installed on the passed global object`);
+  }
+
+  const obj = Object.create(ctor.prototype);
+  Object.defineProperty(obj, utils.implSymbol, {
+    value: Object.create(implModule.implementation.prototype),
+    configurable: true
+  });
+  obj[utils.implSymbol][utils.wrapperSymbol] = obj;
+  return obj[utils.implSymbol];
+};
+
+const originalPromise = Promise;
+const originalPromiseThen = Promise.prototype.then;
+const originalPromiseResolve = Promise.resolve;
+const originalPromiseReject = Promise.reject;
+
+const promiseSideTable = new WeakMap();
+
+// A specialization of https://heycam.github.io/webidl/#invoke-a-callback-function
+// Can be replaced when https://github.com/jsdom/webidl2js/pull/123 lands.
+exports.promiseInvoke = (func, args, thisArg) => {
+  args = args.map(utils.tryWrapperForImpl);
+  try {
+    return Promise.resolve(Reflect.apply(func, thisArg, args));
+  } catch (e) {
+    return Promise.reject(e);
+  }
+};
+
+// A specialization of https://heycam.github.io/webidl/#invoke-a-callback-function
+// Can be replaced when https://github.com/jsdom/webidl2js/pull/123 lands.
+exports.invoke = (func, args, thisArg) => {
+  args = args.map(utils.tryWrapperForImpl);
+  return Reflect.apply(func, thisArg, args);
+};
+
+// https://heycam.github.io/webidl/#a-new-promise
+function newPromise() {
+  // The stateIsPending tracking only works if we never resolve the promises with other promises.
+  // In this spec, that happens to be true for the promises in question; they are always resolved with undefined.
+  const sideData = { stateIsPending: true };
+  const promise = new originalPromise((resolve, reject) => {
+    sideData.resolve = resolve;
+    sideData.reject = reject;
+  });
+
+  promiseSideTable.set(promise, sideData);
+  return promise;
+}
+
+// https://heycam.github.io/webidl/#resolve
+function resolvePromise(p, value) {
+  promiseSideTable.get(p).resolve(value);
+  promiseSideTable.get(p).stateIsPending = false;
+}
+
+// https://heycam.github.io/webidl/#reject
+function rejectPromise(p, reason) {
+  promiseSideTable.get(p).reject(reason);
+  promiseSideTable.get(p).stateIsPending = false;
+}
+
+// https://heycam.github.io/webidl/#a-promise-resolved-with
+function promiseResolvedWith(value) {
+  // Cannot use original Promise.resolve since that will return value itself sometimes, unlike Web IDL.
```

Comment:
Is this observable? Would it change the timing of resolving _startPromise_?

Reply:
It is, because then we don't end up with a weak map entry for the promise sometimes.

This implementation is actually more in line with Web IDL's "promised resolved with" than one that uses `Promise.resolve`; Web IDL always creates a new promise.